src/envelope.rs

```diff
@@ -21,47 +21,31 @@ pub trait MessageEnvelope: Send {
     /// The type of actor that this envelope carries a message for
     type Actor;
 
-    /// Handle the message inside of the box by calling the relevant `AsyncHandler::handle` or
-    /// `Handler::handle` method, returning its result over a return channel if applicable. The
-    /// reason that this returns a future is so that we can propagate any `Handler` responder
-    /// futures upwards and `.await` on them in the manager loop. This also takes `Box<Self>` as the
+    /// Starts the instrumentation of this message request. This will create the request span.
+    fn start_span(&mut self, msg_name: &'static str);
```

Comment:
Is this really the best way we can implement this?

From reading the PR description "create span on first poll", I did not expect something like this but a change to `poll` of `SendFuture`. I haven't tried implementing this myself so perhaps this is impossible but I am finding the indirection here quite confusing.

Also not really happy with the amount of conditional code that we are growing here.

Reply:
> From reading the PR description "create span on first poll", I did not expect something like this but a change to poll of SendFuture. I haven't tried implementing this myself so perhaps this is impossible but I am finding the indirection here quite confusing.

I thought about this but wasn't entirely sure what the best design would be for this so I prototyped it this way (part of the work was getting the semantics right about when the span should actually be created).

One early concept I had was storing the message in `SendFuture` in a new `New` variant of the inner type. However, this refers to the actor type, so I wasn't sure whether this would work with the `sending_erased` version.

> Also not really happy with the amount of conditional code that we are growing here.

Are you referring to feature-conditional/cfg code or branches?

Reply:
It would be nice if we could flatten out the layers in such a way that we can start the instrumentation at a point where we still now the message and actor types but are still outside the `SendFuture`.

> > From reading the PR description "create span on first poll", I did not expect something like this but a change to poll of SendFuture. I haven't tried implementing this myself so perhaps this is impossible but I am finding the indirection here quite confusing.
> 
> I thought about this but wasn't entirely sure what the best design would be for this so I prototyped it this way (part of the work was getting the semantics right about when the span should actually be created).
> 
> One early concept I had was storing the message in `SendFuture` in a new `New` variant of the inner type. However, this refers to the actor type, so I wasn't sure whether this would work with the `sending_erased` version.
> 
> > Also not really happy with the amount of conditional code that we are growing here.
> 
> Are you referring to feature-conditional/cfg code or branches?

Feature-conditional code.